README.markdown
# remixr

The Ruby BestBuy [Remix API](http://remix.bestbuy.com/docs) gem. Remix is an API that gives you access to BestBuy.com's product catalog data and more. 

## Install

    sudo gem install remixr
 
## Usage

    Remixr.api_key = 'OU812' # get yours from http://remix.bestbuy.com/apps/register
    client = Remixr::Client.new
 
### Find stores

    # find stores within 50 miles of ZIP 76227
    stores = client.stores({:area => ['76227', 50]}).fetch.stores

    stores.first

    => {"city"=>"Denton", "longName"=>"Best Buy - Denton", "name"=>"Denton", "region"=>"TX", "address"=>"1800 S Loop 288, Ste 102 Bldg 1 ", "country"=>"US", "lng"=>-97.10067, "postalCode"=>"76205", "phone"=>"940-384-9581", "hours"=>"Mon: 10-9; Tue: 10-9; Wed: 10-9; Thurs: 10-9; Fri: 10-10; Sat: 10-10; Sun: 11-8", "storeId"=>827, "fullPostalCode"=>"76205", "lat"=>33.192524, "distance"=>9.79}
 
 
### Find products

    # fetch first page of products on sale below 20 bucks
    products = client.products({:salePrice => {'$lt' => 20.00}}).fetch.products

    # fetch only SKU and salePrice 
    products = client.products({:salePrice => {'$lt' => 20.00}}).fetch.products
    
### Chaining

You can also chain `stores` and `products` to return stores and nested product info or vice versa

    # find stores within 50 miles of ZIP 76227 and products over three G's
    
    stores = client.stores({:area => ['76227', 50]}).products({:salePrice => {'$gt' => 3000}}).fetch.stores
    
    stores.first.products.first.shortDescription
    
    #=> "ENERGY STAR Qualified 4 HDMI inputs; gray Touch of Color bezel; 16:9 aspect ratio"

### Fetching

All calls terminate in a call to `fetch` which takes the following options

    :page - positive integer for page number
    :show - comma delimited string or array of field names to show
    :sort - hash or string of sort info {'fieldname' => 'asc|desc'}


### Conditional operators

We took a page out of [MongoDB's](http://www.mongodb.org/display/DOCS/Advanced+Queries#AdvancedQueries-ConditionalOperators%3A%3C%2C%3C%3D%2C%3E%2C%3E%3D) playbook and mapped conditional operators to text equivalents to avoid having these be keys to hashes:

    {:salePrice => {'$gte' => 300.00}}
    => salePrice > 300.00
    
    $gte - greater than or equal to : field > value
    $lte - less than or equal to : field > value
    $gt - greater than : field > value
    $lt - less than : field > value
    $ne - not equal to : field != value

More in the examples folder:

[http://github.com/squeejee/remixr/tree/master/examples](http://github.com/squeejee/remixr/tree/master/examples)

### Documentation

[http://rdoc.info/projects/squeejee/remixr](http://rdoc.info/projects/squeejee/remixr)

#### Copyright

Copyright (c) 2009 Squeejee. See LICENSE for details.


History
0.0.2 - updated for ruby 1.9.3
0.0.1 - initial release

test/test_helper.rb
require 'rubygems'
require 'test/unit'
require 'shoulda'
require 'matchy'
require 'mocha'
require 'fakeweb'

# need rails for our tests to run
gem 'actionpack', '>= 3.0.0'
gem 'activesupport', '>= 3.0.0'
gem 'activemodel', '>= 3.0.0'
gem 'railties', '>= 3.0.0'

require 'action_controller/railtie'


FakeWeb.allow_net_connect = false

$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
$LOAD_PATH.unshift(File.dirname(__FILE__))
require 'remixr'

class Test::Unit::TestCase
end

def fixture_file(filename)
  return '' if filename == ''
  file_path = File.expand_path(File.dirname(__FILE__) + '/fixtures/' + filename)
  File.read(file_path)
end

def remix_url(url)
  url =~ /^http/ ? url : "http://api.remix.bestbuy.com:80/v1#{url}"
end

def stub_get(url, filename, status=nil)
  options = {:body => fixture_file(filename)}
  options.merge!({:status => status}) unless status.nil?
  
  FakeWeb.register_uri(:get, remix_url(url), options)
end

def stub_post(url, filename)
  FakeWeb.register_uri(:post, remix_url(url), :body => fixture_file(filename))
end


test/remixr/client_test.rb
require_relative '../test_helper.rb'

class ClientTest < Test::Unit::TestCase
  include Remixr
  
  context "Searching for stores" do
    setup { @client = Remixr::Client.new('OU812') }
    
    should "find all stores" do
      stub_get '/stores?apiKey=OU812&format=json', 'stores.json'
      stores = @client.stores.fetch.stores
      stores.size.should == 10
      stores.first.name.should == 'Caguas'
    end
    
    should "find stores by postal code as option" do
      stub_get '/stores(postalCode=02184)?apiKey=OU812&format=json', 'stores_by_zip_02184.json'
      
      stores = @client.stores({'postalCode' => '02184'}).fetch.stores
      stores.size.should == 1
      stores.first.name.should == 'Braintree'
    end
    
    should "find stores in zip" do
      stub_get '/stores(postalCode=02184)?apiKey=OU812&format=json', 'stores_by_zip_02184.json'
      
      stores = @client.stores.in_zip('02184').fetch.stores
      stores.size.should == 1
      stores.first.name.should == 'Braintree'
    end
    
    should "find stores starting with zip" do
      stub_get '/stores(postalCode=021*)?apiKey=OU812&format=json', 'stores_starting_with_zip_021.json'
      
      stores = @client.stores.in_zip('021*').fetch.stores
      stores.size.should == 5
      stores.first.name.should == 'Back Bay'
    end
    
    should "find stores in a region" do
      stub_get '/stores(region=TX)?apiKey=OU812&format=json', 'stores_in_texas.json'
      
      stores = @client.stores.in_region('TX').fetch.stores
      stores.size.should == 10
      stores.first.name.should == 'Lake Charles'
    end
    
    should "find stores within a radius of a postal code" do
      stub_get '/stores(area(76227,50))?apiKey=OU812&format=json', 'stores_within_50_of_76227.json'
      
      stores = @client.stores({'area' => [76227,50]}).fetch.stores
      stores.size.should == 10
      stores.first.name.should == 'Denton'
    end
    
    should "find stores matching a given list of store IDs" do
      stub_get '/stores(storeId%20in(307,281,612))?apiKey=OU812&format=json', 'stores_in_list.json'
      
      stores = @client.stores({'storeId' => {'$in' => [307,281,612]}}).fetch.stores
      stores.size.should == 3
      stores.first.name.should == 'Richfield'
    end
    
  end
  
  context "Searching for products" do
    setup { @client = Remixr::Client.new('OU812') }
    
    should "find all products" do
      stub_get '/products?apiKey=OU812&format=json', 'products.json'
      products = @client.products.fetch.products
      products.size.should == 10
      products.first.sku.should == 7000032
    end
    
    
    should "find all products with sale price greater than twenty dollars" do
      stub_get '/products(salePrice%3E20)?apiKey=OU812&format=json', 'products_over_20_bucks.json'
      products = @client.products({'salePrice' => {'$gt' => 20}}).fetch.products
      products.size.should == 10
      products.first.sku.should == 16306454
    end
    
    should "find all products with based on search string" do
      stub_get '/products(search=digital%20photo%20printer)?apiKey=OU812&format=json', 'product_search_digital_photo_printer.json'
      products = @client.products({'search' => "digital photo printer"}).fetch.products
      products.size.should == 2
      products.first.sku.should == 4660432
    end  
    
  end
  
  context "When chaining product and store searches" do
    setup { @client = Remixr::Client.new('OU812') }
    
    should "return a products with nested stores" do
      stub_get '/products(salePrice%3E3000)+stores(area(76227,50))?apiKey=OU812&format=json', 'products_over_3k_within_50_miles.json'
      results = @client.products({'salePrice' => {'$gt' => 3000}}).stores({'area' => ['76227', 50]}).fetch
      results.should_respond_to(:products)
      results.should_not_respond_to(:stores)
      results.products.first.sku.should == 9234198
    end
  end
  
  context "When chaining store and product searches" do
    setup { @client = Remixr::Client.new('OU812') }
    
    should "return a products with nested stores" do
      stub_get '/stores(area(76227,50))+products(salePrice%3E3000)?apiKey=OU812&format=json', 'stores_within_50_miles_products_over_3k.json'
      results = @client.stores({'area' => ['76227', 50]}).products({'salePrice' => {'$gt' => 3000}}).fetch
      results.stores.first.name.should == 'Denton'
      results.stores.first.products.first.sku.should == 9234198
    end
  end
  
  context "When using the API" do
    
    setup { @client = Remixr::Client.new('OU812') }
    
    should "support paging" do
      stub_get '/stores?apiKey=OU812&format=json&page=20', 'stores_page_20.json'
      stores = @client.stores.fetch({:page => 20}).stores
      stores.size.should == 10
      stores.first.storeId.should == 282
    end
    
    should "support sorting" do
      stub_get '/stores?apiKey=OU812&format=json&page=20&sort=storeId.desc', 'stores_page_20_sort_by_store_id_desc.json'
      stores = @client.stores.fetch({:page => 20, :sort => {'storeId' => 'desc'}}).stores
      stores.size.should == 10
      stores.first.storeId.should == 1159
    end
    
    should "support limiting return fields" do
      stub_get '/products?apiKey=OU812&format=json&show=sku', 'products_skus_only.json'
      products = @client.products.fetch({:show => 'sku'}).products
      products.size.should == 10
      products.first.sku.should == 7000032
      products.first.format.should == nil
    end
    
  end
end

test/remixr/stores_test.rb
require File.dirname(__FILE__) + '/../test_helper'

class StoresTest < Test::Unit::TestCase
  include Remixr
  
  context "Getting all stores" do
    should "work" do
      stub_get '/v1/stores?apiKey=OU812&format=json', 'stores.json'
      Remixr.api_key = 'OU812'
      stores = Remixr::Stores.all
      stores.size.should == 10
      stores.first.name.should == 'Caguas'
    end
    
  end
  
  context "Getting stores by zip" do
    should "work" do
      stub_get '/v1/stores(postalCode=02184)?apiKey=OU812&format=json', 'stores_by_zip_02184.json'
      Remixr.api_key = 'OU812'
      stores = Remixr::Stores.by_zip("02184")
      stores.size.should == 1
      stores.first.name.should == 'Braintree'
    end
    
  end
end

examples/find_products.rb
Remixr.api_key = 'OU812' # get yours from http://remix.bestbuy.com/apps/register
client = Remixr::Client.new

# fetch first page of products on sale below 20 bucks
products = client.products({:salePrice => {'$lt' => 20.00}}).fetch.products

products.first
# => {"url"=>"http://www.bestbuy.com/site/olspage.jsp?skuId=7524468&type=product&id=1494684&cmp=RMX&ky=1uYnAnk1pohGgQVwEClo8PHv4SN4C5MeL", "inStoreAvailability"=>true, "albumLabel"=>nil, "genre"=>nil, "sku"=>7524468, "affiliateUrl"=>nil, "cjAffiliateAddToCartUrl"=>nil, "salesRankLongTerm"=>2451, "addToCartUrl"=>"http://www.bestbuy.com/site/olspage.jsp?id=pcmcat152200050035&type=category&cmp=RMX&ky=1uYnAnk1pohGgQVwEClo8PHv4SN4C5MeL&qvsids=7524468", "remoteControlImage"=>nil, "affiliateAddToCartUrl"=>nil, "largeImage"=>nil, "new"=>false, "salePrice"=>15.99, "artistId"=>"BYPart1061717", "inStoreAvailabilityUpdateDate"=>"2007-07-27T20:20:50", "largeFrontImage"=>nil, "department"=>"VIDEO/COMPACT DISC", "artistName"=>"Purple City", "accessoriesImage"=>nil, "format"=>"CD", "customerReviewCount"=>nil, "regularPrice"=>15.99, "amgId"=>nil, "freeShipping"=>false, "salesRankShortTerm"=>250, "shortDescription"=>nil, "specialOrder"=>false, "nationalFeatured"=>false, "manufacturer"=>nil, "salesRankMediumTerm"=>570, "energyGuideImage"=>nil, "inStoreAvailabilityText"=>"Store Pickup: Limited Availability", "alternateViewsImage"=>nil, "mobileUrl"=>"http://m.bestbuy.com/r/1494684/7524468/", "subclass"=>"RAP", "name"=>"\"Paris to Purple City\"", "cjAffiliateUrl"=>nil, "image"=>"http://images.bestbuy.com/BestBuy_US/images/products/7524/7524468.jpg", "shippingCost"=>1.59, "subclassId"=>1009, "upc"=>"823979901426", "onlineAvailabilityText"=>"Shipping: Usually leaves our warehouse in 1 business day", "onlineAvailabilityUpdateDate"=>"2007-07-27T20:20:50", "topViewImage"=>nil, "class"=>"COMPACT DISC", "active"=>true, "classId"=>77, "parentalAdvisory"=>false, "thumbnailImage"=>"http://images.bestbuy.com/BestBuy_US/images/products/7524/7524468s.jpg", "leftViewImage"=>nil, "rightViewImage"=>nil, "productId"=>1494684, "source"=>"bestbuy", "customerReviewAverage"=>nil, "activeUpdateDate"=>"2007-07-27T20:20:50", "mediumImage"=>nil, "startDate"=>#<Date: 4906945/2,0,2299161>, "departmentId"=>8, "angleImage"=>nil, "albumVersion"=>nil, "spin360Url"=>nil, "priceUpdateDate"=>"2008-02-09T00:10:37", "categoryPath"=>[<Mash id="cat00000" name="Best Buy">, <Mash id="abcat0600000" name="Music & Movies">, <Mash id="cat02001" name="Music">, <Mash id="cat02012" name="Rap & Hip-Hop">, <Mash id="cat02719" name="Rap">], "printOnly"=>false, "releaseDate"=>#<Date: 4907365/2,0,2299161>, "search"=>nil, "copyright"=>nil, "onSale"=>false, "itemUpdateDate"=>"2009-02-03T03:50:13", "albumTitle"=>"\"Paris to Purple City\"", "type"=>"Music", "originalReleaseDat


# fetch only SKU and salePrice 
products = client.products({:salePrice => {'$lt' => 20.00}}).fetch.products
# => [{"salePrice"=>10.99, "sku"=>7000032}, {"salePrice"=>10.99, "sku"=>11467058}, {"salePrice"=>13.99, "sku"=>7873884}, {"salePrice"=>11.99, "sku"=>15444798}, {"salePrice"=>13.99, "sku"=>9113256}, {"salePrice"=>13.99, "sku"=>5797649}, {"salePrice"=>13.99, "sku"=>8890961}, {"salePrice"=>14.99, "sku"=>17338506}, {"salePrice"=>11.99, "sku"=>9242642}, {"salePrice"=>13.99, "sku"=>9171353}]

lib/remixr/client.rb
module Remixr
  class Client
    include HTTParty
    base_uri 'api.remix.bestbuy.com/v1'
    format :json
    
    attr_reader :api_key, :store_filters, :product_filters, :category_filters
    
    # Get your api_key found here http://remix.bestbuy.com/apps/register
    def initialize(api_key=nil)
      @api_key = api_key
      @api_key ||= Remixr.api_key
      
      @api_path = ''
      @store_filters = {}
      @product_filters = {}
      @category_filters = {}
    end
    
    # Example filters:
    # Stores within 50 miles of ZIP 76227
    #   stores({:area => [76227,50]})
    #   GET /v1/stores(area(76227,50))
    #  
    # Stores west of the Pecos
    #   stores({:lng => {'$lt' => -104.304199}})
    #   GET /v1/stores(lng>-104.304199)
    #
    def stores(filters={})
      unless @api_path.include?('stores()')
        @api_path += '+' unless @api_path.blank?
        @api_path += "stores()"
      end
      @store_filters.merge!(filters)
      self
    end
    
    # Convenience method for finding a store by zip
    #   stores.in_zip(76227)
    #   GET /v1/stores(postalCode=76227)
    def in_zip(zip)
      self.stores({'postalCode' => zip})
    end
    
    # Convenience method for finding a store by region
    #   stores.in_region('TX')
    #   GET /v1/stores(region=TX)
    def in_region(region)
      self.stores({'region' => region})
    end
    
    # Example filters:
    # Products under 20 bucks
    #   products({:salePrice => {'$lt' => 20.00}})
    #   GET /v1/products(salePrice<20.00)
    def products(filters={})
      unless @api_path.include?('products()')
        @api_path += '+' unless @api_path.blank?
        @api_path += "products()"
      end
      @product_filters.merge!(filters)
      self
    end
    
    # Example filters:
    # Products in Gift Center
    #   categories({:name => "Gift Center"})
    #   GET /v1/categorie(name=”Gift Center”)
    def categories(filters={})
      unless @api_path.include?('categories()')
        @api_path += '+' unless @api_path.blank?
        @api_path += "categories()"
      end
      @category_filters.merge!(filters)
      self
    end
    
    
    # Executes the search
    # Possible options:
    # :page - positive integer for page number
    # :show - comma delimited string or array of field names to show
    # :sort - hash or string of sort info {'fieldname' => 'asc|desc'}
    def fetch(options={})
      opts = {:apiKey => @api_key, :format => 'json'}
      opts.merge!(scrub_options(options))
      
      apply_store_filters
      apply_product_filters
      apply_category_filters
      @api_path = URI.escape(@api_path)
      response = self.class.get("/" + @api_path, :query => opts)
      @api_path = ''
      Mash.new response
    end
    
    private
      def apply_store_filters
        if @store_filters.keys.blank? 
          @api_path.gsub!("stores()", "stores")
        else
          @api_path.gsub!("stores()", "stores(#{filter_params_string(@store_filters)})")
        end
      end
      
      def apply_product_filters
        if @product_filters.keys.blank? 
          @api_path.gsub!("products()", "products")
        else
          @api_path.gsub!("products()", "products(#{filter_params_string(@product_filters)})")
        end
      end
      
      def apply_category_filters
        if @category_filters.keys.blank? 
          @api_path.gsub!("categories()", "categories")
        else
          @api_path.gsub!("categories()", "categories(#{filter_params_string(@category_filters)})")
        end
      end
      
      def filter_params_string(filters)
        return "" unless filters.is_a?(Hash)
        
        filters = Mash.new filters.to_hash # calling Mash.new on a mash is bad mmmm k?
        params = []
        filters.each do |key, value|
          if value.is_a?(Hash)
            inner_hash = value
            case inner_hash.keys.first
            when "$gt"
              params << "#{key}>#{inner_hash.values.first}"
            when "$lt"
              params << "#{key}<#{inner_hash.values.first}"
            when "$gte"
              params << "#{key}>=#{inner_hash.values.first}"
            when "$lte"
              params << "#{key}<=#{inner_hash.values.first}"
            when "$ne"
              params << "#{key}!=#{inner_hash.values.first}"
            when "$in"
              params << "#{key} in(#{inner_hash.values.first.join(',')})"
            end
          elsif value.is_a?(Array)
            params << "#{key}(#{value.join(',')})"
          else
            params << "#{key}=#{value}"
          end
        end
        
        if params.blank?
          ""
        else
          "#{params.join('&')}"
        end
      end
      
      def scrub_options(options)
        options = Mash.new(options.to_hash)
        show = options.delete('show')
        unless show.blank?
          if show.is_a?(String)
            options['show'] = show
          elsif show.is_a?(Array)
            options['show'] = show.join(',')
          end
        end
        sort = options.delete('sort')
        unless sort.blank?
          if sort.is_a?(Hash)
            options['sort'] = "#{sort.keys.first}.#{sort.values.first}"
          else
            options['sort'] = sort
          end
        end
        options
      end
    
  end
end